### PR TITLE
Add AMI build job to build workflow

### PR DESCRIPTION
**Pull request by jsf9k**
## 🗣 Description ##

This pull request adds a `build` task to the `build` workflow.  This new task goes through the process of building a new AMI but does not actually register it.

This pull request is blocked until we move to Packer 1.7 (see #70 and cisagov/setup-env-github-action#32), since [that is where support for the `skip_create_ami` key was added for Packer configurations](https://github.com/hashicorp/packer/blob/master/CHANGELOG.md#improvements-3) (see the first bullet).

## 💭 Motivation and context ##

This is a useful feature to add since it will catch changes that cause the AMI build process to fail.  Currently these changes are only caught when we create a pre-release or release.

N.B.:  Once we have dev, staging, and production COOL environments we could allow this job to go ahead and register the AMI in the dev environment.  That would allow us to fully test in the dev environment before we create a pre-release (for staging) and/or a release (for production).

## 🧪 Testing ##

cisagov/setup-env-github-action#32 was merged and now all `pre-commit` hooks and GitHub Actions workflows pass. 

## ✅ Checklist ##

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] Tests have been added and/or modified to cover the changes in this PR.
* [x] All new and existing tests pass.
* [x] Add `build/build` job to the list of GH Actions jobs that must pass before a PR can be merged.

### CODE REVIEW

**Review by felddy — APPROVED** [review]
This is a good improvement.  It should catch a whole new class of errors.
Strong work. 💪

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -133,3 +136,95 @@ jobs:
...
+            ${{ env.CURL_CACHE_DIR }}/"${TERRAFORM_ZIP}"
+          sudo ln -s /opt/terraform/terraform /usr/bin/terraform
+          sudo mv /usr/local/bin/terraform /usr/local/bin/terraform-default
+          sudo ln -s /opt/terraform/terraform /usr/local/bin/terraform
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install --upgrade \
+                      --requirement requirements.txt
+      - name: Install ansible roles
```
Comment (dav3r) [inline .github/workflows/build.yml]:
```suggestion
      - name: Install Ansible roles
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -133,3 +136,95 @@ jobs:
...
+      - name: Install Terraform
+        run: |
+          TERRAFORM_ZIP="terraform_${TERRAFORM_VERSION}_linux_amd64.zip"
+          curl --output ${{ env.CURL_CACHE_DIR }}/"${TERRAFORM_ZIP}" \
+            --time-cond ${{ env.CURL_CACHE_DIR }}/"${TERRAFORM_ZIP}" \
+            --location \
+            "https://releases.hashicorp.com/terraform/${TERRAFORM_VERSION}/${TERRAFORM_ZIP}"
+          sudo unzip -d /opt/terraform \
+            ${{ env.CURL_CACHE_DIR }}/"${TERRAFORM_ZIP}"
+          sudo ln -s /opt/terraform/terraform /usr/bin/terraform
```
Comment (mcdonnnj) [inline .github/workflows/build.yml]: Mirroring the decision in https://github.com/cisagov/skeleton-generic/pull/82#discussion_r667214249

```suggestion
```

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -133,3 +136,95 @@ jobs:
...
+      - name: Install Packer
+        run: |
+          mkdir -p ${{ env.CURL_CACHE_DIR }}
+          PACKER_ZIP="packer_${PACKER_VERSION}_linux_amd64.zip"
+          curl --output ${{ env.CURL_CACHE_DIR }}/"${PACKER_ZIP}" \
+            --time-cond ${{ env.CURL_CACHE_DIR }}/"${PACKER_ZIP}" \
+            --location \
+            "https://releases.hashicorp.com/packer/${PACKER_VERSION}/${PACKER_ZIP}"
+          sudo unzip -o -d /usr/local/bin \
+            ${{ env.CURL_CACHE_DIR }}/"${PACKER_ZIP}"
```
Comment (mcdonnnj) [inline .github/workflows/build.yml]: Mirroring https://github.com/cisagov/skeleton-generic/pull/82#discussion_r667207173

```suggestion
          sudo unzip -d /opt/packer \
            ${{ env.CURL_CACHE_DIR }}/"${PACKER_ZIP}"
          sudo mv /usr/local/bin/packer /usr/local/bin/packer-default
          sudo ln -s /opt/packer/packer /usr/local/bin/packer
```

File: .github/workflows/build.yml (hunk 1 above)
Comment (mcdonnnj) [inline .github/workflows/build.yml]:
```suggestion
      - name: Install Ansible roles
```

**Review by dav3r — APPROVED** [review]
Excellent! 🎸 
I only noted one small thang.

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -133,3 +136,95 @@ jobs:
...
+            tf-${{ env.TERRAFORM_VERSION }}-"
+        with:
+          path: |
+            ${{ env.PIP_CACHE_DIR }}
+            ${{ env.CURL_CACHE_DIR }}
+          key: "${{ env.BASE_CACHE_KEY }}\
+            ${{ hashFiles('**/requirements.txt') }}"
+          restore-keys: |
+            ${{ env.BASE_CACHE_KEY }}
+      - name: Install Packer
```
Comment (mcdonnnj) [inline .github/workflows/build.yml:191]: Mirroring https://github.com/cisagov/skeleton-generic/pull/82/files#r667205518

```suggestion
      - name: Setup curl cache
        run: mkdir -p ${{ env.CURL_CACHE_DIR }}
      - name: Install Packer
```

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -133,3 +136,95 @@ jobs:
...
+        run: |
+          mkdir -p ${{ env.CURL_CACHE_DIR }}
```
Comment (mcdonnnj) [inline .github/workflows/build.yml]:
```suggestion
```

**Review by mcdonnnj — CHANGES_REQUESTED** [review]
This is a great improvement. I have some suggestions that are mostly to bring this into alignment with the work in https://github.com/cisagov/skeleton-generic/pull/82.

File: .github/workflows/build.yml (hunk 1 above)
Comment (jsf9k) [reply .github/workflows/build.yml]: See commit e95501d.

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -133,3 +136,97 @@ jobs:
       - name: Setup tmate debug session
         uses: mxschmitt/action-tmate@v3
         if: env.RUN_TMATE
+  build:
+    # The AMI build process is an expensive test (in terms of time) so
+    # let's not run it unless the other jobs succeed.
+    needs:
+      - lint
+      - test
```
Comment (mcdonnnj) [inline .github/workflows/build.yml]: This is how it's done in [the cisagov/skeleton-python-library build job](https://github.com/cisagov/skeleton-python-library/blob/4ef2364701c0db074c423f740062b2d7c197bea8/.github/workflows/build.yml#L113). I'm fine if we want to switch to this style, but I'd like to use the same format for similar functionality whichever way we decide.

```suggestion
    needs: [lint, test]
```

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -133,3 +136,97 @@ jobs:
...
+          aws-region: ${{ env.BUILD_REGION }}
+          role-to-assume: ${{ secrets.BUILD_ROLE_TO_ASSUME_STAGING }}
+          role-duration-seconds: 3600
+      # When called by Packer, Ansible will find /usr/bin/python3 and
+      # use it; therefore, we must ensure that /usr/bin/python3 points
+      # to the version of Python that we installed in the
+      # actions/setup-python step above.  This can hose other tasks
+      # that are expecting to find the system Python at that location,
+      # though, so we undo this change after running Packer.
+      - name: Create a /usr/bin/python3 symlink to the installed python
```
Comment (mcdonnnj) [inline .github/workflows/build.yml]:
```suggestion
      - name: Create a /usr/bin/python3 symlink to the installed Python
```

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -133,3 +136,97 @@ jobs:
...
+        env:
+          GITHUB_IS_PRERELEASE: ${{ github.event.release.prerelease }}
+          GITHUB_RELEASE_TAG: ${{ github.event.release.tag_name }}
+          GITHUB_RELEASE_URL: ${{ github.event.release.html_url }}
+        # This runs through the AMI creation process but does not
+        # actually create an AMI
+        run: |
+          packer build --timestamp-ui -var skip_create_ami=true \
+          src/packer.json
+      - name: Remove /usr/bin/python3 symlink to the installed python
```
Comment (mcdonnnj) [inline .github/workflows/build.yml]:
```suggestion
      - name: Remove /usr/bin/python3 symlink to the installed Python
```

**Review by mcdonnnj — APPROVED** [review]
just some last suggestions. Thanks again for thinking of and implementing this improvement!

**Pull request merged**